Applications/Spire/Source/Ui/OpenFilterPanel.cpp

```diff
@@ -0,0 +1,377 @@
+  m_list_view = new ListView(m_matches,
+    std::bind_front(&OpenFilterPanel::make_item, this));
+  m_list_view->get_current()->connect_update_signal(
+    std::bind_front(&OpenFilterPanel::on_list_view_current, this));
+  m_matches_connection = m_matches->connect_operation_signal(
+    std::bind_front(&OpenFilterPanel::on_matches_operation, this));
+  auto scrollable_list_box = new ScrollableListBox(*m_list_view);
+  scrollable_list_box->setMinimumSize(scale_width(164), scale_height(80));
+  scrollable_list_box->setMaximumHeight(scale_height(158));
+  layout->addWidget(scrollable_list_box);
+  m_filter_panel = new FilterPanel(std::move(title), this, parent);
+  m_filter_panel->connect_reset_signal(
+    std::bind_front(&OpenFilterPanel::on_reset, this));
+  auto panel_window = window();
+  panel_window->setWindowFlags(Qt::Tool | panel_window->windowFlags());
+  panel_window->installEventFilter(this);
+}
+
+const std::shared_ptr<AnyListModel>& OpenFilterPanel::get_matches() const {
+  return m_matches;
+}
+
+const std::shared_ptr<ValueModel<OpenFilterPanel::Mode>>&
+    OpenFilterPanel::get_mode() const {
+  return m_mode;
+}
+
+connection OpenFilterPanel::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+bool OpenFilterPanel::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::Close) {
+    m_filter_panel->hide();
+    hide();
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+bool OpenFilterPanel::event(QEvent* event) {
+  if(event->type() == QEvent::ShowToParent) {
+    m_filter_panel->show();
+    window()->activateWindow();
+    m_input_box->setFocus();
+  } else if(event->type() == QEvent::HideToParent) {
+    m_filter_panel->hide();
+  }
+  return QWidget::event(event);
+}
+
+QWidget* OpenFilterPanel::make_item(const std::shared_ptr<AnyListModel>& model,
+    int index) {
+  auto label = displayTextAny(model->get(index));
+  auto item = new DeletableItem(label);
+  item->connect_delete_signal([=] {
+    auto index = [&] {
+      for(auto i = 0; i < m_matches->get_size(); ++i) {
+        if(label == displayTextAny(model->get(i))) {
+          return i;
+        }
+      }
+      return -1;
+    }();
+    if(index >= 0) {
+      m_matches->remove(index);
+      m_input_box->setFocus();
+    }
+  });
+  return item;
+}
+
+void OpenFilterPanel::on_input_box_submission(const AnyRef& submission) {
+  m_input_box->get_current()->set({});
+  m_matches->push(submission);
+}
+
+void OpenFilterPanel::on_list_view_current(const optional<int>& current) {
+  m_list_view->setFocusPolicy(Qt::NoFocus);
+}
+
+void OpenFilterPanel::on_mode_current(Mode mode) {
+  submit();
+}
+
+void OpenFilterPanel::on_matches_operation(
+    const AnyListModel::Operation& operation) {
+  visit(operation,
+    [&] (const AnyListModel::AddOperation& operation) {
+      auto item = m_list_view->get_list_item(operation.m_index);
+      set_style(*item, LIST_ITEM_STYLE());
```

Comment:
Why is the style being set here on every add operation?

Reply:
I tried to set the style of the `ListItem` in the `ListView`, but the style couldn’t be applied to the newly added `ListItem` through the `ListView`.

Reply:
I found a solution to solve it.